### ISSUE TITLE
Querying display "safe area" (notch / cutouts / rounded corners)

### ISSUE THREAD

**Issue by PrimaryFeather** [airsdk/Adobe-Runtime-Support#933]
Today's mobile displays come in a variety of shapes, with notches, cutouts, rounded corners, etc. Mobile OS add status bars of varying sizes and "home indicator" bars, which might be visible or not, depending on the "full screen" display mode. And finally, you might rotate your app at runtime, having those elements be moved to different sides of the screen.

In short, there's a lot to manage – and AIR doesn't provide any information about the "safe area" of the current view. In fact, I'd go so far to say that you _cannot_ create a mobile app nowadays without relying on an ANE like the [excellent one from distriqt](https://docs.airnativeextensions.com/docs/application/display/cutouts). 

Really, I just tried to get the "Starling Scaffold Project" to handle this – and I wasn't able to pull it off. On Android, it's kind of possible, because the drawable display area starts below a possible notch, and stops before a possible home indicator (downside is that you cannot customize status bar colors). On iPhone, the status bar is simply drawn on top of the stage – which gives a lot of flexibility, but means you really need to know what's there. One could add some guesswork like "if the aspect ratio >= 2.0, there's a notch" – but even that fails when you want to support iPad. The only way would be to hard-code the device resolutions that have a home button, vs. those which use a home indicator. It's just not feasible. 

So, as much as I'm happy for distriqt that they have a really fantastic USP for their "Application" ANE (kudos to @marchbold!), I think it's high time that there is at least an API to query this information, or maybe even rethinking the available display modes to make implementing this easier. (The former would probably be a much lower hanging fruit. 😄)

Thanks in advance for considering it!

**Comment by marchbold** [comment]
Thanks, must say the Application ANE is the first one I add to an app. I would love to be able to provide it for free but it is one of the extensions that has required the most work, working out how the display modes change with AIR has taken a lot of time.  

I definitely agree some of the functionality there should be part of AIR, I'll have a chat to Andrew if there's something we can arrange here.

**Comment by PrimaryFeather** [comment]
I totally understand that you're not able to provide this for free, Michael! I can only imagine how much work must gone into this ANE, and it really belongs into any AIR project nowadays, IMHO. :smile:

I'm really just asking about a way to forward the "safe area" info from the native SDKs, because not even a simple demo project for AIR can be set up without it.

**Comment by MalacTheLittle** [comment]
@PrimaryFeather I totally agree with you that some basic info should be present in AIR itself.
Actually I suggest to include it to `Screen` class since it's really usefull on Windows and Mac OS when getting `bounds` and `visibleBounds` for `mainScreen`. 
Since you can add support for android notch and fullscreen in apps decriptor it would be perfect if `visibleBounds` would show height of notch (or `y` of `visibleBounds`) and height of Navigaton bar (or `height` of `visibleBounds`).

@marchbold Is there any progress with your arrangement?

**Comment by marchbold** [comment]
@MalacTheLittle I totally forgot about this, will add it to my list.

**Comment by TerryCavanagh** [comment]
Hi there!

I'm updating an old app, and just wanted query the progress of this: is it still the case that there is no way in December 2022 to query the cutout position without the use of an ANE?

**Comment by PrimaryFeather** [comment]
Actually, this was added with AIR-50! 🎉

There's now a [safeArea property](https://airsdk.dev/reference/actionscript/3.0/flash/display/Screen.html#safeArea) on the Screen class. 

I haven't tried it out myself yet, but this should be what you're looking for, Terry.
In other words, this probably means I can close this issue! Yay! Thanks to everybody who was involved. 😄👏

**Comment by TerryCavanagh** [comment]
Thank you!

You should maybe *not* close this issue, though, as I had a lot of trouble trying to find an answer to this question, and this issue is the closest thing I could find on google! If you close this issue, you'll leave future googlers stuck 😅

**Comment by PrimaryFeather** [comment]
Good point – maybe it should be left open until this information becomes more common knowledge. 😉

**Comment by TerryCavanagh** [comment]
Thanks! (and also just to confirm, this new property works great and completely fixed my problem)

**Comment by mtilhan** [comment]
I tested it on iPhone 11 pro and it works. The notch issue fixed for me. @PrimaryFeather thanks for bringing out this problem, and thanks to all Harman team for this feature :)

**Comment by Mintonist** [comment]
Hi all!

With latest AIR 51 do we have API to get info about "safe area" (notch / cutouts / rounded corners) ?
At least fact that it present on device.
Thx!

**Comment by ajwfrost** [comment]
Hi @Mintonist  - if you mean https://airsdk.dev/reference/actionscript/3.0/flash/display/Screen.html#safeArea then yes?
If there are issues with this please let us know - I think I recall that it only really applies if you've got an app displayed in full-screen (as otherwise, the normal display area (i.e. visible bounds) are fully within the safe area anyway).
thanks

**Comment by Mintonist** [comment]
>  I think I recall that it only really applies if you've got an app displayed in full-screen

You mean
```
<initialWindow>      
        <fullScreen>true</fullScreen>   
</initialWindow>
```
?
